Uyanda.Coffee.Persistence/Accessors/BeverageAccessor.cs

```diff
@@ -65,61 +67,190 @@ public async Task<InvoiceModel> PurchaseAsync(IEnumerable<LineItemModel> lineIte
+
+                customer.Points += earnedPoints;
 
-                beveragePrice.Beverage.Name = price.Beverage.Name;
+                await localDbContext.AddAsync(ToEntity(customer));
 
-                beveragePrice.Cost = price.Cost;
+                await localDbContext.SaveChangesAsync();
 
-                beveragePrice.Beverage.IsActive = price.Beverage.IsActive;
+                var customerInvoice = new InvoiceModel
+                {
+                    Date = DateTime.Now,
+                    LineItems = purchase,
+                    CustomerId = customer.Id
+                };
 
-                beveragePrice.BeverageSizeId = price.BeverageSizeId;
+                await localDbContext.AddAsync(ToEntity(customerInvoice));
 
                 await localDbContext.SaveChangesAsync();
 
-                return ToModel(beveragePrice);
+                return customerInvoice;
             }
 
+            throw new InvalidOperationException("Invalid input");
 
-            if(price.BeverageId == 0)
-            {
-                await localDbContext.BeverageCost.AddAsync(ToEntity(price));
+        }
+
+        private async Task<InvoiceModel> DiscountPurchaseAsync(IEnumerable<LineItemModel> lineItems, CustomerModel customer)
+        {
+            var isUserFound = await localDbContext.Customer
+                .Where(c => c.Id == customer.Id)
+                .AnyAsync();
+
+            if (isUserFound == false)
+                throw new InvalidOperationException("Phone number does not exist");
+
+            var beveragePrices = await localDbContext.BeverageCost.AsNoTracking()
+                .Select(c => new { c.Id, c.Cost })
+                .ToDictionaryAsync(item => item.Id, item => item.Cost);
+
+            var totalCost = lineItems
+                .Sum(c => c.Count * beveragePrices[c.Id]);
+
+            var customerEntity = await localDbContext.Customer
+                .Where(c => c.Id == customer.Id)
+                .SingleAsync();
+
+            var discount = customerEntity.Points;
 
-                localDbContext.SaveChanges();
+            totalCost -= discount;
 
-                return price;
+            if (totalCost < 0)
+            {
+                discount = - totalCost;
+
+                totalCost = 0;
+            }
+            else
+            {
+                discount = 0;
             }
 
-            var isBeverageAvailable = await localDbContext.BeverageCost.AsNoTracking()
-                .Where(c => c.BeverageId == price.BeverageId)
-                .AnyAsync();
+            customerEntity.Points = discount;
+
+            await localDbContext.SaveChangesAsync();
 
-            if(isBeverageAvailable)
+            var beverageCost = new InvoiceEntity
             {
-                price.Beverage = null;
+                Date = DateTime.Now,
+                LineItems = lineItems.Select(ToEntity).ToList(),
+                CustomerId = customer.Id
+            };
```

Comment:
Please store _IsRedeemingPoints_, _DiscountedPoints_ and _FinalInvoicePrice_ from in the DB.